feat(framework): Add FedRDF robust dynamic aggregation strategy

francordel opened the pull request:
## Issue

Closes [#6043](https://github.com/adap/flower/issues/6043).

---

## Description

Flower currently lacks a **robust aggregation strategy** to defend against
**poisoning attacks** in Federated Learning. When malicious clients send
corrupted model updates, standard FedAvg aggregation can be heavily compromised,
leading to a significant degradation in global model performance.

---

## Related issues/PRs

- Enables the new baseline proposed in #6039.

---

## Proposal

Implement **FedRDF (Robust and Dynamic Aggregation Function)**, a strategy that
defends against poisoning attacks through adaptive, frequency-domain aggregation.

### Explanation

FedRDF introduces the following:

1. **Adaptive switching:** detects poisoning using a Kolmogorov-Smirnov test on
   the skewness of client updates.
2. **Frequency-domain robust aggregation:** per coordinate, the sorted client
   weights are projected into the frequency domain to estimate their density
   function (Nanbu, 1995), and the client weight exhibiting the highest
   estimated density is selected (paper Eq. 11, Sec. V). The selection depends
   on the actual weight values, is order/sign-invariant, and excludes weights
   isolated from the high-density (benign) regions, i.e., malicious updates.
3. **Configurable threshold:** the `threshold` parameter controls the
   robustness-efficiency trade-off.
   - `threshold <= 0`: always use the frequency-domain robust aggregation.
   - `threshold > 0`: switch adaptively based on the detected skewness
     (K-S detection requires >= 10 replies per round; documented in the docstring).
4. **Layer-wise processing:** robust aggregation is applied to each model layer
   individually for fine-grained resilience.

**Implementation details**
- Inherits from `FedAvg` (under `serverapp/strategy`).
- Implements `aggregate_train()` using the new `Message` / `*Record` API
  (`Iterable[Message]`, `ArrayRecord`, `MetricRecord`) instead of `Fit*Ins/Res`.
- The FedAvg path delegates to the shared `aggregate_arrayrecords` helper, so
  weighting behaves exactly like the other serverapp strategies.
- The frequency-domain path processes coordinates in fixed-size chunks to bound
  peak memory on large models.
- Fully typed with `NDArray` (passes `mypy --strict`), vectorized NumPy
  throughout, per-instance RNG (`np.random.default_rng()`).
- `scipy` is an **optional** dependency: it is imported behind `try/except` and
  only required for adaptive mode (`threshold > 0`), where the K-S test runs. The
  constructor raises a clear `ImportError` only when adaptive mode is requested
  without scipy; `threshold <= 0` runs without it.

---

## Testing

- 11 unit tests, all passing locally (`serverapp/strategy/fedrdf_test.py`).
- Coverage includes: the always-FFT path (`threshold <= 0`), the weighted-FedAvg
  path (`threshold > 0`, low skewness), the adaptive switch to the
  frequency-domain path on high skewness, the K-S / skewness helpers, and the
  scipy-optional contract (construction with and without scipy; scipy-dependent
  tests are skipped via a `requires_scipy` marker, verified: 11 passed with
  scipy, 6 passed / 5 skipped without).
- Poison-rejection tests confirm that both a **+100x** and a **-100x** poisoned
  client are excluded from the aggregate (distance to benign clients is far
  smaller than distance to the poisoned update).
- Full quantitative evaluation of the paper's experiments is provided separately
  by the baseline in #6039.

---

## Reference

E. Mármol Campos, A. González-Vidal, J. L. Hernández-Ramos, and A. Skarmeta,
*"FedRDF: A Robust and Dynamic Aggregation Function Against Poisoning Attacks in
Federated Learning"*, *IEEE Transactions on Emerging Topics in Computing*,
vol. 13, no. 1, pp. 48-67, 2025.
DOI: [10.1109/TETC.2024.3474484](https://doi.org/10.1109/TETC.2024.3474484) ·
[arXiv](https://arxiv.org/abs/2402.10082)

---

## Checklist

- [x] Implement proposed change.
- [x] Write tests (11 unit tests, all passing locally).
- [ ] Update documentation (no public API docs page included, happy to add one).
- [ ] Make CI checks pass (workflows show `action_required`, they need a
      maintainer to approve the runs; locally `format.sh` and the Python checks
      in `test.sh` pass: isort/black/ruff/docsig/mypy/pylint).
- [x] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`).

---

## Any other comments?

**Files changed (3):**
- `framework/py/flwr/serverapp/strategy/fedrdf.py` (309 lines), strategy implementation.
- `framework/py/flwr/serverapp/strategy/fedrdf_test.py` (293 lines), test suite (11 tests).
- `framework/py/flwr/serverapp/strategy/__init__.py`, added `FedRDF` export.

The legacy `framework/py/flwr/server/strategy` package is no longer touched by
this PR (the earlier accidental edit was reverted, so it is back in sync with
`main` and out of the diff).

francordel committed:
feat(framework): add FedRDF robust dynamic aggregation strategy

francordel committed:
Merge branch 'main' of https://github.com/adap/flower into feat/fedrdf-strategy

francordel committed:
Merge branch 'main' of https://github.com/adap/flower into feat/fedrdf-strategy

francordel commented:
Hi @WilliamLindskog,
yes, I’m still on it. I had to pause for a few weeks due to other deadlines, but I’m back now. I’m addressing Javier’s feedback (moving FedRDF to the new ServerApp/Message-based pattern without initialize_parameters, handling the scipy import, and tightening the tests). I’ll push updates and respond to the open comments today or in the next few days. Thanks for your patience.

francordel committed:
Merge branch 'main' of https://github.com/adap/flower into feat/fedrdf-strategy

francordel committed:
feat(strategy): Migrate FedRDF to serverapp/strategy with Message API

Addresses review feedback from PR #6044 by @danieljanes:
"We recently migrated all strategies to serverapp/strategy and make use
of the Message abstraction instead of FitIns/Res etc. Could you update
the strategy in that way? Also, because scipy is not a package that comes
with Flower, we'll need to put the import in a try/except."

Migration implementation:
- Moved from server/strategy to serverapp/strategy
- Replaced FitIns/FitRes with Message abstraction (Iterable[Message])
- Use ArrayRecord/MetricRecord instead of Parameters/Metrics
- Updated aggregate_train/aggregate_evaluate signatures for Message API
- Wrapped scipy import in try/except with HAS_SCIPY flag (lines 31-37)
- Added ImportError in __init__ with installation instructions (lines 133-137)
- Migrated all unit tests to new API

Performance optimization:
During migration testing, discovered critical bottleneck in _compute_skewness
that made threshold > 0 functionality impractical:
- Sample max 100 weight positions instead of testing all positions
- Reduces K-S test complexity from O(n_parameters * 100) to O(100 * 100)
- Enables practical use of adaptive FedAvg/FFT switching
- Threshold=0.5 now completes in ~97s vs hanging indefinitely

Scipy dependency handling:
Import wrapped in try/except block with HAS_SCIPY flag. Strategy __init__
raises ImportError with clear message if scipy not installed, directing
users to install scipy>=1.7.0.

Files changed:
- framework/py/flwr/serverapp/strategy/fedrdf.py (new, migrated + optimized)
- framework/py/flwr/serverapp/strategy/fedrdf_test.py (new, migrated)
- framework/py/flwr/serverapp/strategy/__init__.py (add FedRDF export)
- framework/py/flwr/server/strategy/__init__.py (remove old export)
- framework/py/flwr/server/strategy/fedrdf.py (deleted)
- framework/py/flwr/server/strategy/fedrdf_test.py (deleted)

Testing:
- All unit tests passing in new location
- Threshold functionality verified working (0.0, 0.5, 1.0 tested)
- ImportError raised correctly when scipy not available

Related: PR #6044

francordel commented:
Good evening!
I have addressed the comments:
feat(strategy): Migrate FedRDF to serverapp/strategy with Message API

Addresses review feedback from PR #6044 by @danieljanes:
"We recently migrated all strategies to serverapp/strategy and make use
of the Message abstraction instead of FitIns/Res etc. Could you update
the strategy in that way? Also, because scipy is not a package that comes
with Flower, we'll need to put the import in a try/except."

Migration implementation:
- Moved from server/strategy to serverapp/strategy
- Replaced FitIns/FitRes with Message abstraction (Iterable[Message])
- Use ArrayRecord/MetricRecord instead of Parameters/Metrics
- Updated aggregate_train/aggregate_evaluate signatures for Message API
- Wrapped scipy import in try/except with HAS_SCIPY flag (lines 31-37)
- Added ImportError in __init__ with installation instructions (lines 133-137)
- Migrated all unit tests to new API

Performance optimization:
During migration testing, discovered critical bottleneck in _compute_skewness
that made threshold > 0 functionality impractical:
- Sample max 100 weight positions instead of testing all positions
- Reduces K-S test complexity from O(n_parameters * 100) to O(100 * 100)
- Enables practical use of adaptive FedAvg/FFT switching
- Threshold=0.5 now completes in ~97s vs hanging indefinitely

Scipy dependency handling:
Import wrapped in try/except block with HAS_SCIPY flag. Strategy __init__
raises ImportError with clear message if scipy not installed, directing
users to install scipy>=1.7.0.

Files changed:
- framework/py/flwr/serverapp/strategy/fedrdf.py (new, migrated + optimized)
- framework/py/flwr/serverapp/strategy/fedrdf_test.py (new, migrated)
- framework/py/flwr/serverapp/strategy/__init__.py (add FedRDF export)
- framework/py/flwr/server/strategy/__init__.py (remove old export)
- framework/py/flwr/server/strategy/fedrdf.py (deleted)
- framework/py/flwr/server/strategy/fedrdf_test.py (deleted)

Testing:
- All unit tests passing in new location
- Threshold functionality verified working (0.0, 0.5, 1.0 tested)
- ImportError raised correctly when scipy not available

WilliamLindskog commented on framework/py/flwr/server/strategy/__init__.py:
```diff
@@ -64,7 +64,7 @@
     "FedMedian",
     "FedOpt",
-    "FedProx",
+    "FedProx"
```
This drops the comma after `"FedProx"`, so Python concatenates it with the next string literal and `__all__` ends up containing `"FedProxFedTrimmedAvg"` instead of separate `FedProx`/`FedTrimmedAvg` exports. Since this PR moved FedRDF to `serverapp/strategy`, this legacy init change should probably be reverted.

francordel replied:
Sorry, that was a mistake I missed, reverted it, comma restored. Thanks for catching it!

WilliamLindskog commented on framework/py/flwr/serverapp/strategy/fedrdf.py:
```diff
@@ -0,0 +1,445 @@
+    ) -> None:
+        """Initialize FedRDF strategy."""
+        # Check scipy dependency
+        if not HAS_SCIPY:
```
This makes `scipy` mandatory when constructing `FedRDF`, even for `threshold <= 0` where the K-S test is never used. Since Flower does not ship `scipy`, could this be lazy-failed only for adaptive mode, or added through an explicit dependency/extra if we want the strategy to require it?

francordel replied:
@WilliamLindskog Good point. Done. `scipy` is now only required in adaptive mode (`threshold > 0`); with `threshold <= 0` it's never imported or needed. The constructor raises a clear `ImportError` only when adaptive mode is requested without scipy, and the import stays behind `try/except`. Added a test covering both cases.

WilliamLindskog commented on framework/py/flwr/serverapp/strategy/fedrdf.py:
```diff
@@ -0,0 +1,445 @@
+
+            # Get weight (num_examples) from metrics
+            weight = 1.0  # Default weight
+            if "metrics" in record_dict:
```
This path still assumes the metric record key is literally `"metrics"`. The ServerApp strategies/utilities usually rely on `RecordDict.metric_records`/`array_records` after validation, so custom metric record keys would silently default every client weight to `1.0` here.

francordel replied:
Fixed. It now reads the ArrayRecord and weighting MetricRecord via `RecordDict.array_records` / `metric_records` and `self.weighted_by_key` (same as `aggregate_arrayrecords`, after `_check_and_log_replies` validates the replies). Dropped the silent `1.0` default, so custom record keys no longer collapse the weights. @WilliamLindskog

francordel replied:
Follow-up: as of 629e55f18 the FedAvg path now delegates directly to
`aggregate_arrayrecords`, so the weighting logic is exactly the shared one used
by the other serverapp strategies.

WilliamLindskog reviewed (commented):
Thanks for moving this to the ServerApp strategy path. I left a few comments on issues that look worth addressing before another maintainer pass. Two broader notes that do not map cleanly to a single code line:

- The PR checklist/body currently says docs were updated and CI passes, but the current diff only touches the strategy implementation, tests, and init exports, and I do not see check runs/status contexts on the current head commit.
- The PR body still describes the older `framework/py/flwr/server/strategy` file layout and says 10 tests, while the current branch adds 7 tests under `serverapp/strategy`. It would help to refresh the PR description after the next push.

WilliamLindskog commented on framework/py/flwr/serverapp/strategy/fedrdf.py:
```diff
@@ -0,0 +1,445 @@
+            dominant_idx = int(np.argmax(magnitudes))
+
+            # Use the original value corresponding to the dominant frequency
+            result.flat[i] = values[dominant_idx]
```
I do not think `dominant_idx` can be used as an index back into the client values. It is the index of an FFT frequency bin, not the client that should be selected. As written, this does not reconstruct via inverse FFT or filter components as described in the PR, and it can return the poisoned value depending on client order/sign.

francordel replied:
@WilliamLindskog You're right, that was a bug. Reworked it to follow the paper's Eq. 11: per coordinate, the client weights are first sorted into the vector V, then projected into the frequency domain, and the weight exhibiting the highest frequency is selected. Sorting first makes the selection order-invariant, so it no longer depends on client order/sign or returns the poisoned value. To be precise, the method *selects* a representative weight in the frequency domain rather than reconstructing via inverse FFT.
 I've updated the PR description, which previously over-described an IFFT step. Also vectorized it, and added a poison-rejection test confirming a 100× outlier client is excluded from the aggregate.

francordel replied:
Update: the latest push (629e55f18) refines this further. `_fourier_aggregate`
now implements the density-based selection described in Sec. V of the paper
(Nanbu, 1995): per coordinate, the sorted client weights are projected into the
frequency domain to estimate their density function, and the client weight
exhibiting the highest estimated density is selected. The selection therefore
depends on the actual weight values (not only on the number of clients), is
order/sign-invariant, and the poison-rejection tests now cover both +100× and
−100× outlier clients.

francordel committed:
Merge remote-tracking branch 'upstream/main' into feat/fedrdf-strategy

francordel committed:
fix(strategy): address FedRDF review comments

- Revert accidental dropped comma after "FedProx" in the legacy
  server/strategy __init__ (keep FedProx/FedTrimmedAvg as separate exports)
- Require scipy only in adaptive mode (threshold > 0); threshold <= 0
  no longer needs scipy as the K-S test is never used
- Read arrays and weights via array_records/metric_records instead of a
  hard-coded "metrics" key, so custom record keys no longer default the weight
- Reconstruct the Fourier-based aggregation per the reference: sort the
  per-coordinate client weights and select the one at the highest frequency
- Vectorize the aggregation and align comments with sibling strategies
- Add tests for scipy-optional construction and the adaptive FFT path

francordel committed:
Merge remote-tracking branch 'upstream/main' into feat/fedrdf-strategy

francordel committed:
refactor(strategy): address automated review feedback

- Import Message/Record types from flwr.app for consistency with sibling
  strategies (keep log from flwr.common)
- Drop the redundant FedRDF re-export alias in the strategy __init__
- Guard _ks_proportion with HAS_SCIPY and sample skewness positions before
  stacking to avoid materializing the full model for large models
- Skip the scipy-dependent tests when scipy is not installed, since scipy is
  an optional dependency only used in adaptive mode

francordel commented on framework/py/flwr/serverapp/strategy/fedrdf.py:
```diff
@@ -0,0 +1,284 @@
+        """
+        stacked = np.sort(np.stack([arr.flatten() for arr in arrays]), axis=0)
+        index = int(np.argmax(np.abs(np.fft.fftfreq(stacked.shape[0]))))
+        return stacked[index].reshape(arrays[0].shape)
```
This follows the reference implementation of the published method (Eq. 11): the per-coordinate client weights are sorted and the one exhibiting the highest frequency is selected, which is what the docstring states. The intent is to faithfully reproduce the paper's aggregation, not to introduce a variant. The poison-rejection test confirms it excludes a 100× outlier client.

francordel commented on framework/py/flwr/serverapp/strategy/fedrdf.py:
```diff
@@ -0,0 +1,284 @@
+    def _ks_proportion(self, sample: np.ndarray) -> float:
+        """Return the proportion of K-S tests that detect divergence in a sample."""
+        if len(sample) < 10:
+            return 0.0
```
Good point — `_compute_skewness` now samples the positions first and only gathers those coordinates (`np.ravel(arr)[sample_indices]`), so it no longer materializes the full model. Fixed.

francordel committed:
fix(strategy): make FedRDF Fourier aggregation value-dependent via density selection.
- Rework _fourier_aggregate: estimate the per-coordinate density of client
  weights through the frequency domain (Nanbu, 1995) and select the weight
  with the highest density, per Eq. 11 of the paper
- Process coordinates in chunks to bound peak memory
- Reuse aggregate_arrayrecords for the FedAvg path and adopt NDArray typing
  (fixes mypy --strict)
- Add poison-rejection test for a negative-scaled (-100x) outlier client
- Document that adaptive K-S detection requires >= 10 replies per round

francordel committed:
Merge branch 'main' of https://github.com/adap/flower into feat/fedrdf-strategy

francordel commented on framework/py/flwr/serverapp/strategy/fedrdf_test.py:
```diff
@@ -0,0 +1,253 @@
+def test_fedrdf_aggregate_train_with_high_threshold() -> None:
+    """Test aggregate_train with high threshold (should use weighted FedAvg)."""
+    # Prepare
+    strategy = FedRDF(threshold=1.0)  # High threshold
```
This test is already guarded: it is decorated with the `requires_scipy` marker
(`pytest.mark.skipif(not HAS_SCIPY, ...)`, defined at the top of the file), so it
is skipped when scipy is not installed rather than failing. Verified locally by
running the suite without scipy: 6 passed, 5 skipped, and with scipy: 11 passed.

francordel commented on framework/py/flwr/serverapp/strategy/fedrdf_test.py:
```diff
@@ -0,0 +1,253 @@
+def test_fedrdf_ks_proportion() -> None:
+    """Test _ks_proportion method."""
+    # Prepare
+    strategy = FedRDF()
```
Same guard applies here — this test is decorated with the `requires_scipy` skipif
marker defined at the top of the file, so it is skipped when scipy is not
installed (no need for `importorskip` inside the test body). Verified locally
without scipy: the suite reports it as skipped.

francordel commented on framework/py/flwr/serverapp/strategy/fedrdf_test.py:
```diff
@@ -0,0 +1,253 @@
+def test_fedrdf_compute_skewness() -> None:
+    """Test _compute_skewness method."""
+    # Prepare
+    strategy = FedRDF()
```
Already covered by the `requires_scipy` marker at the current head, so it is
skipped when scipy is unavailable instead of forcing the optional dependency.
The scipy-optional contract itself is tested explicitly in
`test_fedrdf_requires_scipy_only_in_adaptive_mode`, which runs with and without
scipy installed.

francordel commented on framework/py/flwr/serverapp/strategy/fedrdf_test.py:
```diff
@@ -0,0 +1,253 @@
+def test_fedrdf_weighted_aggregation() -> None:
+    """Test FedRDF with different weights (num_examples)."""
+    # Prepare
+    strategy = FedRDF(threshold=1.0)  # High threshold to use weighted FedAvg
```
This one is also decorated with `requires_scipy`, so it does not force the
optional dependency, in an environment without scipy it is reported as skipped
rather than failing the suite.

francordel commented on framework/py/flwr/serverapp/strategy/fedrdf_test.py:
```diff
@@ -0,0 +1,253 @@
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    """Test that high skewness routes aggregation through the FFT path."""
+    strategy = FedRDF(threshold=0.5)
```
Guarded with `requires_scipy` as well — adaptive-mode tests only run when scipy
is present, so the suite stays green in minimal environments (verified locally:
without scipy this test is skipped, not failed).

francordel reviewed (commented):
Replied to the automated review comments. Most of them reference the previous
commit in the range, the `requires_scipy` markers and the `ImportError` guard
they suggest were already in place at head 022a07ca. The latest push goes
further and reworks `_fourier_aggregate` into the value-dependent, density-based
selection described in Sec. V of the paper (629e55f18), adds a negative-scaled
(−100×) poison-rejection test (11 tests total), and adopts NDArray typing to
pass mypy --strict.

francordel commented on framework/py/flwr/serverapp/strategy/__init__.py:
```diff
@@ -30,6 +30,7 @@
 from .fedavgm import FedAvgM
 from .fedmedian import FedMedian
 from .fedprox import FedProx
+from .fedrdf import FedRDF as FedRDF
```
Fixed since 022a07ca the export now uses the direct form
`from .fedrdf import FedRDF`, consistent with the rest of the module (the
thread shows as outdated for that reason).

francordel commented on framework/py/flwr/serverapp/strategy/fedrdf.py:
```diff
@@ -0,0 +1,284 @@
+except ImportError:
+    HAS_SCIPY = False
+
+from flwr.common import Array, ArrayRecord, Message, MetricRecord, RecordDict, log
```
Aligned since 022a07ca — `Array`, `ArrayRecord`, `Message`, `MetricRecord` and
`RecordDict` are imported from `flwr.app`, with only `log` coming from
`flwr.common`, matching `fedavg.py` and the other serverapp strategies.

francordel commented on framework/py/flwr/serverapp/strategy/fedrdf.py:
```diff
@@ -0,0 +1,284 @@
+            sample_indices = np.arange(n_positions)
+
+        proportions = [self._ks_proportion(standardized[:, i]) for i in sample_indices]
+        return float(np.mean(proportions))
```
Fixed since 022a07ca — the sample positions are drawn first and only those
coordinates are gathered (`np.ravel(arr)[sample_indices]`), so the full model is
never flattened/stacked. The latest push (629e55f18) also bounds peak memory on
the Fourier aggregation path by processing coordinates in fixed-size chunks.

WilliamLindskog committed:
Merge branch 'main' into feat/fedrdf-strategy

francordel committed:
Merge branch 'main' into feat/fedrdf-strategy

francordel committed:
Merge branch 'main' into feat/fedrdf-strategy
